README.md

```diff
@@ -2,22 +2,118 @@
+const contract = new ethers.Contract(
+  MINER_TO_PEERID_CONTRACT_ADDRESS,
+  MINER_TO_PEERID_CONTRACT_ABI,
+  provider,
+)
+
+// Basic usage
+const minerId = 'f0142637'
+const { peerId, source } = await getIndexProviderPeerId(minerId, contract)
+console.log(peerId) // e.g., '12D3KooWMsPmAA65yHAHgbxgh7CPkEctJHZMeM3rAvoW8CZKxtpG'
+console.log(source) // 'smartContract' or 'filecoinMinerInfo'
+
+// Advanced usage with options
+const { peerId, source } = await getIndexProviderPeerId(minerId, contract, {
+  maxAttempts: 3, // Number of retry attempts (default: 5)
+  rpcUrl: 'https://custom-filecoin-api.com', // Custom Filecoin RPC endpoint (default: 'https://api.node.glif.io/')
+  rpcAuth: 'your-auth-token', // Optional authorization token for RPC
+  rpcFn: customRpcFunction, // Optional custom RPC function implementation.
+})
+```
+
+The function prioritizes the smart contract result. If the smart contract doesn't return a valid
+PeerId, it falls back to the result from FilecoinMinerInfo. If both methods fail, an error is thrown
+with details about the failure.
+
+### Parameters
+
+- `minerId`: A Filecoin miner actor ID (e.g., `f0142637`)
+- `smartContract`: An ethers.js Contract instance for the Filecoin smart contract
+- `options`: (Optional) Configuration object:
+  - `maxAttempts`: Number of retry attempts for RPC calls (default: 5)
+  - `rpcUrl`: Filecoin RPC endpoint URL (default: 'https://api.node.glif.io/')
+  - `rpcAuth`: Authorization token for RPC endpoint (if required)
+  - `rpcFn`: Custom function for making RPC calls.
+
+#### Custom RPC Function
+
+- `rpcFn`: Custom function for making RPC calls to the Filecoin network. If provided, it overrides
+  the default RPC implementation. This function should match the following signature:
+
+  ```typescript
+  async function rpcFn(
+    method: string,
+    params: any[],
+    options?: {
+      rpcUrl?: string
+      rpcAuth?: string
+    },
+  ): Promise<any>
+  ```
+
+  The function receives:
+
+  - `method`: The RPC method name to call
+  - `params`: An array of parameters for the method
+  - `options`: Optional object containing:
+    - `rpcUrl`: The Filecoin RPC endpoint URL
+    - `rpcAuth`: Authorization token for the RPC endpoint
+
+  It should return a Promise that resolves to the RPC response's result field.
+
+  If not provided, the default RPC function will be used with the specified `rpcUrl` and `rpcAuth`.
+  The default implementation uses 'https://api.node.glif.io/' as the RPC endpoint if no `rpcUrl` is
+  provided and it uses no authentication (rpcAuth is undefined).
+
+### Return Value
+
+Returns a Promise that resolves to the miner's PeerId string (e.g.,
+`12D3KooWMsPmAA65yHAHgbxgh7CPkEctJHZMeM3rAvoW8CZKxtpG`).
+
+### Constants
+
+#### Smart Contract Integration
 
 This library interacts with the MinerPeerIDMapping smart contract on the Filecoin EVM to retrieve
 peer ID data.
 
-#### `MINER_TO_PEERID_CONTRACT_ADDRESS`
+##### `MINER_TO_PEERID_CONTRACT_ADDRESS`
```

Comment:
AFAIK, you can only nest headings to four levels (h4 in HTML).

You can use bold font instead of headings if you need deeper nesting.

```suggestion
**`MINER_TO_PEERID_CONTRACT_ADDRESS`**
```

Reply:
Alternatively we rethink the nesting. I think individual constants don't need their own heading, this can just be a list:

```
- `MINER_TO_..`: Description
- etc
```

Reply:
I guess if GitHub can render level-5 headings, then it's okay to use them. 🤷🏻